### PR TITLE
[WIP] Unity master staging timezone fallback

### PR DESCRIPTION
**Pull request by joncham**

### CODE REVIEW

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+#if UNITY
```
Comment (joshpeterson) [inline mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: Should we use `UNITY` or `PLATFORM_UNITY` for this code? I think in the C++ code we used `PLATFORM_UNITY`. We don't need to use the same define, but it is worth considering.

Comment (joncham) [reply mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: We already define `UNITY` at the managed Makefile level

Comment (joshpeterson) [reply mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: Ok

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+			TimeSpan utcOffsetTS = TimeSpan.FromTicks(data[(int)TimeZoneData.UtcOffsetIdx]);
+		    char utcOffsetSign = (utcOffsetTS >= TimeSpan.Zero) ? '+' : '-';
+		    string displayName = "(GMT" + utcOffsetSign + utcOffsetTS.ToString(@"hh\:mm") + ") Local Time";
+		    string standardDisplayName = names[(int)TimeZoneNames.StandardNameIdx];
+		    string daylightDisplayName = names[(int)TimeZoneNames.DaylightNameIdx];
+		    
+		    //Create The Adjustment Rules For This TimeZoneInfo.
+		    var adjustmentList = new List<TimeZoneInfo.AdjustmentRule>();
+		    for(int year = 1970; year <= 2037; year++)
```
Comment (mderoy) [inline mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: You should change this to 1973. the android mktime barfs before then

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+				if (dltStartTime == dltEndTime)
+					continue;
+
+				TimeZoneInfo.TransitionTime startTime = TimeZoneInfo.TransitionTime.CreateFixedDateRule(dltStartTime, dlt.Start.Month, dlt.Start.Day);
+				TimeZoneInfo.TransitionTime endTime = TimeZoneInfo.TransitionTime.CreateFixedDateRule(dltEndTime, dlt.End.Month, dlt.End.Day);
+				
+
+				//mktime only supports dates starting in 1970, so create an adjustment rule for years before 1970 following 1970s rules 
+				if (year == 1970)
```
Comment (mderoy) [inline mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: 1973 too

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs (hunk 2 above)
Comment (mderoy) [reply mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: Also we need to make sure that IOS doesn't have some similar weird issue. 1970 forward worked fine on desktop...

File: mcs/class/corlib/System/TimeZoneInfo.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-
+
```
Comment (mderoy) [inline mcs/class/corlib/System/TimeZoneInfo.cs]: whitespace

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		    DaylightNameIdx
+		};
+		static TimeZoneInfo CreateLocalUnity ()
+		{
+			Int64[] data;
+		    string[] names;
+		    if (!System.CurrentSystemTimeZone.GetTimeZoneData (1970, out data, out names))
+				throw new NotSupportedException ("Can't get timezone name.");
+
+			TimeSpan utcOffsetTS = TimeSpan.FromTicks(data[(int)TimeZoneData.UtcOffsetIdx]);
```
Comment (mderoy) [inline mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: bad indention

File: mcs/class/corlib/System/TimeZoneInfo.cs [hunk 6]
```diff
@@ -165,6 +165,12 @@ static TimeZoneInfo CreateLocal ()
 			}
 #endif
 
+#if UNITY
+			var localUnity = CreateLocalUnity ();
+			if (localUnity != null)
+				return localUnity;
```
Comment (mderoy) [inline mcs/class/corlib/System/TimeZoneInfo.cs]: Do we really want this to kick in on all non-windows systems? we shouldn't return this here, and instead return it if  return FindSystemTimeZoneByFileName (tz, Path.Combine (TimeZoneDirectory, tz)); throws

File: mcs/class/corlib/System/TimeZoneInfo.Unity.cs (hunk 2 above)
Comment (mderoy) [reply mcs/class/corlib/System/TimeZoneInfo.Unity.cs]: We should probably add a basic runtime test that verifys that utc != localtime on all our platforms

**Pull request merged**